### ISSUE TITLE
Invoking `mock_s3` twice leaks key

### ISSUE THREAD

**Issue by hannes-ucsc** [getmoto/moto#5930]
Continued from #5913

```python
$ python -Wall <<END
from boto3 import client
from moto import mock_s3, __version__

print(__version__)

with mock_s3():
    s3 = client('s3')
    s3.create_bucket(Bucket='foo')
    s3.put_object(Bucket='foo', Key='bar', Body='stuff')

with mock_s3():
    pass
END
```

produces

```
4.1.3.dev22
/Users/hannes/workspace/hca/azul/.venv/lib/python3.9/site-packages/moto/s3/models.py:349: ResourceWarning: S3 key was not disposed of in time
  warnings.warn("S3 key was not disposed of in time", ResourceWarning)
```
  
 Didn't look into the cause this time. It must be a regression introduced by PR #5920 because I can't reproduce on 4.1.2.

**Comment by bblommers** [comment]
Thanks for raising this @hannes-ucsc - that was a regression indeed. Should be fixed as of moto >= 4.1.3.dev29

**Comment by hannes-ucsc** [comment]
Thank you for fixing it, @bblommers. I did find another one: https://github.com/getmoto/moto/issues/5937
